### ISSUE TITLE
Plugin CanViewPermission - paginatedPermissionCheck() and listPermissionCheck() SECURITY VULNERABILITY

### ISSUE THREAD

**Issue by Lightnicus** [silverstripe/silverstripe-graphql#553]
We've noticed what we believe is a fairly major security vulnerability when relying on the built in permission check for DataObjects when returning a DataList.

The algorithm in paginatedPermissionCheck() checks canView() on each DataObject in the DataList to be returned if a permission to view exists and if not it collects the IDs of the DataObjects in an array to be excluded in the sql via the exclude() ORM function. This leads to an SQL query assembled by the ORM that excludes the IDs of those DataObjects - and that's where the problem starts. This SQL excludes these IDs but does not make sure that the query doesn't return other datasets in that table.

(simplified) example:
original query created by the ORM:
SELECT fields from Member limit 10 offset 0;

above permission check creates an exclude list for all 10 records returned by above query, resulting in:

SELECT fields from Member WHERE ID NOT IN (1,2,3...10) limit 10 offset 0;

assuming the Member table has more than 20 records, above query will still return 10 DataObjects (just the next 10 instead of the 10 that are excluded)

To make things worse, the function paginatedPermissionCheck() then only checks if the count of expected datasets (by the original query) maches the count of the query that was created using the exclude list. If the count is the same, the original ORM object / query is used to return the results. This behavior complete bypasses the security measure attempted (even if the query with the exclude list would be used, you'd still get DataSets back that should not be returned). 

Possible solution: 
the method listPermissionCheck() needs to include all ID's in the ORM query before excluding the ones that fail canView() - that way the generated SQL query will only ever return the DataObjects that are actually checked by canView()
something like this:
            $originalIDList = $obj->map('ID', 'ID')->toArray();
            $obj = $obj->filter(['ID' => $origIDList]);
            return $obj->exclude(['ID' => $excludes]);

tldr; to summarize, do NOT currently rely on the canViewPermission plugin to provide security for DataList queries on your Model classes.
